org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java

```diff
@@ -116,22 +116,30 @@ public void generateCode(BlockScope currentScope, CodeStream codeStream, boolean
 		codeStream.addVariable(this.secretExpressionValue);
 	}
 	codeStream.instance_of(this.type, this.type.resolvedType);
-	if (this.elementVariable != null) {
-		BranchLabel actionLabel = new BranchLabel(codeStream);
-		codeStream.dup();
-		codeStream.ifeq(actionLabel);
+	if (this.pattern != null) {
+		BranchLabel falseLabel = new BranchLabel(codeStream);
+		BranchLabel trueLabel = new BranchLabel(codeStream);
+		BranchLabel continueLabel = new BranchLabel(codeStream);
+		codeStream.ifeq(falseLabel);
+
 		if (this.secretExpressionValue != null) {
 			codeStream.load(this.secretExpressionValue);
 			codeStream.removeVariable(this.secretExpressionValue);
 		} else {
 			this.expression.generateCode(currentScope, codeStream, true);
 		}
-		codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
-		this.elementVariable.binding.recordInitializationStartPC(codeStream.position);
-		codeStream.store(this.elementVariable.binding, false);
-		codeStream.removeVariable(this.elementVariable.binding);
-		codeStream.recordPositionsFrom(codeStream.position, this.sourceEnd);
-		actionLabel.place();
+		this.pattern.generateOptimizedBoolean(currentScope, codeStream, trueLabel, falseLabel);
+
+		trueLabel.place();
+		codeStream.iconst_1();
```

Comment:
codeStream.dup() in the original code should have been kept instead of const_1() and const_0() [later]. However, not a critical change.

Reply:
> codeStream.dup() in the original code should have been kept instead of const_1() and const_0() [later]. However, not a critical change.

I will follow up on this along with various other cleanups I have called out in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2016#issuecomment-1942909619